drivers/linstorvolumemanager.py

```diff
@@ -196,18 +200,56 @@ def _get_controller_uri():
         # Not found, maybe we are trying to create the SR...
         pass
 
-def get_controller_uri():
-    retries = 0
-    while True:
+
+def get_cached_controller_uri():
+    try:
+        with open(CONTROLLER_CACHE_PATH, "r") as f:
+            return f.read().strip()
+    except Exception as e:
+        util.SMlog('Unable to read controller uri cache file at `{}` : {}'.format(
+            CONTROLLER_CACHE_PATH,
+            e
+        ), priority=util.LOG_DEBUG)
+    return None
+
+def delete_controller_uri_cache():
+    try:
+        os.remove(CONTROLLER_CACHE_PATH)
+    except Exception as e:
+        util.SMlog('Unable to write controller uri cache file at `{}` : {}'.format(
+            CONTROLLER_CACHE_PATH,
+            e
+        ), priority=util.LOG_DEBUG)
+
+def write_controller_uri_cache(uri):
+    try:
+        if not os.path.exists(CONTROLLER_CACHE_DIRECTORY):
+            os.makedirs(CONTROLLER_CACHE_DIRECTORY)
+            os.chmod(CONTROLLER_CACHE_DIRECTORY, 0o700)
+        with open(CONTROLLER_CACHE_PATH, "w") as f:
+            f.write(uri)
+    except Exception as e:
+        util.SMlog('Unable to write controller uri cache file at `{}` : {}'.format(
+            CONTROLLER_CACHE_PATH,
+            e
+        ), priority=util.LOG_DEBUG)
+
+
+def build_controller_uri_cache():
+    for retries in range(10):
         uri = _get_controller_uri()
         if uri:
+            write_controller_uri_cache(uri)
             return uri
+        if retries < 9:
+            time.sleep(1)
+    return None
```

Comment:
I prefer my previous "retry-loop" version for one reason: here we use "range(10)" and "retries < 9". So the values ​​are hard-coded and if we want to change the number of retries, we now have to change two places.
However we we can use another way I guess:
```suggestion
    uri = _get_controller_uri()
    if not uri:
        for retries in range(9):
            time.sleep(1)
            uri = _get_controller_uri()
            if uri:
                break

    if uri:
        write_controller_uri_cache(uri)
    return uri
```

Reply:
I can keep your previous version if you want. It was an oversight on my end indeed. I think I prefer your previous version rather than the one you just proposed, will revert this change if you're okay with that.

Reply:
The previous version may be simpler in its logic but introduces a variable. The second duplicates a line but uses a for loop over a range, which limits coding errors.

I don't have a preference; choose what you want :)

Reply:
I added your suggestion in [8581b70](https://github.com/xcp-ng/sm/pull/83/commits/8581b7092aa4150d1c70aba76bd3411f3c45fb53)